### PR TITLE
5646 item direction table

### LINKED ISSUE #5646: Create & sync Item Directions Table
image

The item_direction table holds default directions for an item
If configured, a drop down is shown of the available directions, and a single direction can be selected - this will add to the prescription
The directions can either be an abbreviation (exact match) or free text

Tasks
- [ ] New Database table in oms `item_direction`
- [ ] ItemDirectionRepositoryRow (find_by_id, delete)
- [ ] Translations for `Upsert` and `Delete`

### PR DESCRIPTION
**Pull request by Chris-Petty**
Fixes #5646 

# 👩🏻‍💻 What does this PR do?

Adds migration, repo and sync logic for the `item direction` table

### CODE REVIEW

File: server/repository/src/db_diesel/item_direction.rs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+use crate::diesel_macros::apply_equal_filter;
+
+use crate::{repository_error::RepositoryError, DBType, EqualFilter};
+use diesel::{dsl::IntoBoxed, prelude::*};
+
+pub type ItemDirection = ItemDirectionRow;
+
+#[derive(Clone, Default, PartialEq, Debug)]
+pub struct ItemDirectionFilter {
+    pub id: Option<EqualFilter<String>>,
```
Comment (jmbrunskill) [inline server/repository/src/db_diesel/item_direction.rs:12]: I think we're going to want a filter on `item_id`, could be done later though, so I'm happy if it's not in this PR.

File: server/repository/src/migrations/v2_05_00/item_direction_create_table.rs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+    fn migrate(&self, connection: &StorageConnection) -> anyhow::Result<()> {
+        sql!(
+            connection,
+            r#"
+            CREATE TABLE item_direction (
+                id TEXT NOT NULL PRIMARY KEY,
+                item_link_id TEXT NOT NULL REFERENCES item_link(id),
+                directions TEXT NOT NULL,
+                priority BIGINT NOT NULL
```
Comment (jmbrunskill) [inline server/repository/src/migrations/v2_05_00/item_direction_create_table.rs:18]: I would have thought `INTEGER` would be ok, I guess big int is fine too.

File: server/service/src/sync/translations/item_direction.rs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    fn try_translate_from_upsert_sync_record(
+        &self,
+        _: &StorageConnection,
+        sync_record: &SyncBufferRow,
+    ) -> Result<PullTranslateResult, anyhow::Error> {
+        let data = serde_json::from_str::<LegacyItemDirectionRow>(&sync_record.data)?;
+
+        let result = ItemDirectionRow {
+            id: data.ID,
+            item_link_id: data.item_ID,
```
Comment (jmbrunskill) [inline server/service/src/sync/translations/item_direction.rs:42]: Is it safe to assume an `item_id` is also an `item_link_id`? Feels like we might need to look up the `item_id` based on this, but then I always get confused with `*_link` tables...

File: server/repository/src/db_diesel/item_direction_row.rs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        directions -> Text,
+        priority -> BigInt,
+    }
+}
+
+#[derive(
+    Clone, Insertable, Queryable, Debug, PartialEq, AsChangeset, Eq, Serialize, Deserialize,
+)]
+#[diesel(table_name = item_direction)]
+#[diesel(treat_none_as_null = true)]
```
Comment (jmbrunskill) [inline server/repository/src/db_diesel/item_direction_row.rs:23]: This line is a little bit pointless with no nullable columns :)

**Review by jmbrunskill — APPROVED** [review]
Thanks Chris, seems to be working well for me!

Couple of comments up to you if you want to address them...

File: server/repository/src/migrations/v2_05_00/item_direction_create_table.rs (hunk 2 above)
Comment (Chris-Petty) [reply server/repository/src/migrations/v2_05_00/item_direction_create_table.rs:18]: I had integer and diesel postgres was throwing a wobbly on i64 not receiving enough bytes

File: server/service/src/sync/translations/item_direction.rs (hunk 3 above)
Comment (Chris-Petty) [reply server/service/src/sync/translations/item_direction.rs:42]: The item link tables are populated with `item.id == item_link.id == item.item_link_id` so it should be fine I think!

File: server/repository/src/db_diesel/item_direction_row.rs (hunk 4 above)
Comment (Chris-Petty) [reply server/repository/src/db_diesel/item_direction_row.rs:23]: Cargo cult 😆

**Pull request merged**